core/store/src/cold_storage.rs

```diff
@@ -0,0 +1,265 @@
+use crate::columns::DBKeyType;
+use crate::refcount::add_positive_refcount;
+use crate::{DBCol, DBTransaction, Database, Store};
+
+use borsh::BorshDeserialize;
+use near_primitives::types::BlockHeight;
+use std::collections::HashMap;
+use std::io;
+use strum::IntoEnumIterator;
+
+type StoreKey = Vec<u8>;
+type StoreValue = Option<Vec<u8>>;
+type StoreCache = HashMap<(DBCol, StoreKey), StoreValue>;
+
+struct StoreWithCache<'a> {
+    store: &'a Store,
+    cache: StoreCache,
+}
+
+pub fn update_cold_db(
+    cold_db: &dyn Database,
+    hot_store: &Store,
+    height: &BlockHeight,
+) -> io::Result<()> {
+    let _span = tracing::debug_span!(target: "store", "update cold db", height = height);
+
+    let mut store_with_cache = StoreWithCache { store: hot_store, cache: StoreCache::new() };
+
+    let key_type_to_keys = get_keys_from_store(&mut store_with_cache, height)?;
+    for col in DBCol::iter() {
+        if col.is_cold() {
+            copy_from_store(
+                cold_db,
+                &mut store_with_cache,
+                col,
+                combine_keys(&key_type_to_keys, &col.key_type()),
+            )?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Gets values for given keys in a column from provided hot_store.
+/// Creates a transaction based on that values with set DBOp s.
+/// Writes that transaction to cold_db.
+fn copy_from_store(
+    cold_db: &dyn Database,
+    hot_store: &mut StoreWithCache,
+    col: DBCol,
+    keys: Vec<StoreKey>,
+) -> io::Result<()> {
+    let _span = tracing::debug_span!(target: "store", "create and write transaction to cold db", col = %col);
+
+    let mut transaction = DBTransaction::new();
+    for key in keys {
+        let data = hot_store.get(col, &key)?;
```

Comment:
```suggestion
        // TODO: Look into using RocksDB’s multi_key function.  It
        // might speed things up.  Currently our Database abstraction
        // doesn’t offer interface for it so that would need to be
        // added.
        let data = hot_store.get(col, &key)?;
```